tests/Unit/Evolution/Systems/Ccz4/Ricci.py

```diff
@@ -0,0 +1,48 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+
+import numpy as np
+
+
+def deriv_christoffel_second_kind(d_conformal_christoffel_second_kind,
+                                  conformal_spatial_metric,
+                                  inverse_conformal_spatial_metric, field_d,
+                                  field_d_up, field_p, d_field_p):
+    return (
+        d_conformal_christoffel_second_kind + 2.0 *
+        (np.einsum("kml,ijl->kmij", field_d_up,
+                   (np.einsum("jl,i", conformal_spatial_metric, field_p) +
+                    np.einsum("il,j", conformal_spatial_metric, field_p) -
+                    np.einsum("ij,l", conformal_spatial_metric, field_p))) -
+         np.einsum("ml,ijkl->kmij", inverse_conformal_spatial_metric,
+                   (np.einsum("kjl,i", field_d, field_p) +
+                    np.einsum("kil,j", field_d, field_p) -
+                    np.einsum("kij,l", field_d, field_p)))) -
+        np.einsum("ml,ijkl->kmij", inverse_conformal_spatial_metric,
+                  (np.einsum("jl,ki", conformal_spatial_metric, d_field_p) +
+                   np.einsum("jl,ik", conformal_spatial_metric, d_field_p) +
+                   np.einsum("il,kj", conformal_spatial_metric, d_field_p) +
+                   np.einsum("il,jk", conformal_spatial_metric, d_field_p) -
+                   np.einsum("ij,kl", conformal_spatial_metric, d_field_p) -
+                   np.einsum("ij,lk", conformal_spatial_metric, d_field_p))) /
+        2.0)
+
+
+def spatial_ricci_tensor(christoffel_second_kind,
+                         d_conformal_christoffel_second_kind,
+                         conformal_spatial_metric,
+                         inverse_conformal_spatial_metric, field_d, field_d_up,
+                         field_p, d_field_p):
+    d_christoffel_second_kind = deriv_christoffel_second_kind(
+        d_conformal_christoffel_second_kind, conformal_spatial_metric,
+        inverse_conformal_spatial_metric, field_d, field_d_up, field_p,
+        d_field_p)
+
+    return np.einsum(
```

Comment:
I don't understand what's going on with einsum here. I was expecting something like `einsum("kmij->mmij", d_christoffel_second_kind) + einsum("jmim", d_crhistoffel_second_kind) +...`

Reply:
This is intended to be the trace of the Riemann tensor (eq. 19 [here](https://arxiv.org/pdf/1707.09910.pdf)) where it's computing `R^m{}_{imj}` since `R_{ij} = R^m{}_{imj}`. The four terms in eq. 19 are ordered `mikj` to align with the ordering for the Riemann tensor, and then the outermost `einsum` computes the trace by using `mimj`.

If this seems correct but was difficult to follow, let me know if you would like me to add comments for this.

If this is incorrect, let me know that, as well!

Reply:
I guess there is also no reason I can't remove the outer `einsum` and just do something like:
```
    return (np.einsum("mmij", d_christoffel_second_kind) -
         np.einsum("jmim", d_christoffel_second_kind) +
         np.einsum("lij,mlm", christoffel_second_kind,
                   christoffel_second_kind) -
         np.einsum("lim,mlj", christoffel_second_kind,
                   christoffel_second_kind))
```

Let me know if you would like me to change it to something like this ^

Reply:
I find the 4 sums much easier to read than what you currently have since it matches what it's in the dox :)